api/v3/Payment.php

```diff
@@ -134,51 +134,72 @@ function civicrm_api3_payment_create(&$params) {
   }
   // Get contribution
   $contribution = civicrm_api3('Contribution', 'getsingle', array('id' => $params['contribution_id']));
-  if ($contribution['contribution_status'] != 'Partially paid') {
-    throw new API_Exception('Please select a contribution which has a partial payment');
+  $contributionStatus = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
+  if ($contributionStatus[$contribution['contribution_status_id']] != 'Partially paid'
+    && !($contributionStatus[$contribution['contribution_status_id']] == 'Pending' && $contribution['is_pay_later'] == TRUE)
+  ) {
+    throw new API_Exception('Please select a contribution which has a partial or pending payment');
   }
   else {
-    $trxn = CRM_Contribute_BAO_Contribution::recordPartialPayment($contribution, $params);
-    $paid = CRM_Core_BAO_FinancialTrxn::getTotalPayments($params['contribution_id']);
-    $total = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $params['contribution_id'], 'total_amount');
-    $cmp = bccomp($total, $paid, 5);
-    if ($cmp == 0 || $cmp == -1) {// If paid amount is greater or equal to total amount
-      civicrm_api3('Contribution', 'completetransaction', array('id' => $contribution['id']));
+    // Check if pending contribution
+    $fullyPaidPayLater = FALSE;
+    if ($contributionStatus[$contribution['contribution_status_id']] == 'Pending') {
+      $cmp = bccomp($contribution['total_amount'], $params['total_amount'], 5);
+      // Total payment amount is the whole amount paid against pending contribution
+      if ($cmp == 0 || $cmp == -1) {
+        civicrm_api3('Contribution', 'completetransaction', array('id' => $contribution['id']));
+        // Get the trxn
+        $trxnId = CRM_Core_BAO_FinancialTrxn::getFinancialTrxnId($contribution['id'], 'DESC');
+        $ftParams = array('id' => $trxnId['financialTrxnId']);
+        $trxn = CRM_Core_BAO_FinancialTrxn::retrieve($ftParams, CRM_Core_DAO::$_nullArray);
+        $fullyPaidPayLater = TRUE;
+      }
+      else {
+        civicrm_api3('Contribution', 'create',
+          array(
+            'id' => $contribution['id'],
+            'contribution_status_id' => array_search('Partially paid', $contributionStatus),
+          )
+        );
+      }
     }
-  }
-  if (CRM_Utils_Array::value('line_item', $params) && !empty($trxn)) {
-    foreach ($params['line_item'] as $values) {
-      foreach ($values as $id => $amount) {
-        $p = array('id' => $id);
-        $check = CRM_Price_BAO_LineItem::retrieve($p, $defaults);
-        if (empty($check)) {
-          throw new API_Exception('Please specify a valid Line Item.');
+    if (!$fullyPaidPayLater) {
+      $trxn = CRM_Core_BAO_FinancialTrxn::getPartialPaymentTrxn($contribution, $params);
+      if (CRM_Utils_Array::value('line_item', $params) && !empty($trxn)) {
+        foreach ($params['line_item'] as $values) {
```

Comment:
ouch!! no default behaviour - always force them to select line items??? That is quite off putting to using this api because I think knowing which line items you want to match is the edge case here